src/lib.rs

```diff
@@ -27,102 +18,783 @@ pub extern "C" fn wapc_init() {
+                spec: Some(apicore::PodSpec {
+                    security_context: Some(apicore::PodSecurityContext {
+                        se_linux_options: Some(expected_selinux_options.clone().into()),
+                        ..apicore::PodSecurityContext::default()
+                    }),
+                    ..pod_spec.clone()
+                }),
+                ..apicore::Pod::default()
+            };
+
+            // If the pod has no SELinuxOptions, we default it to the expected value. We still have
+            // to check that all containers meet the SELinux requirements set by the configuration
+            // because individual containers might override the default PodSecurityContext SELinux
+            // options
+
+            let mut has_to_mutate = false;
+            let pod_selinux_options = if let Some(pod_security_context) = pod_spec.security_context
+            {
+                if let Some(selinux_options) = pod_security_context.se_linux_options {
+                    selinux_options
+                } else {
+                    has_to_mutate = true;
+                    expected_selinux_options.clone().into()
+                }
             } else {
-                info!(LOG_DRAIN, "accepting resource");
-                kubewarden::accept_request()
+                has_to_mutate = true;
+                expected_selinux_options.clone().into()
+            };
+
+            let all_compliant_containers = pod_spec.containers.into_iter().all(|container| {
+                is_selinux_compliant(
+                    &container.security_context,
+                    expected_selinux_options.clone(),
+                    &pod_selinux_options,
+                )
+            });
+
+            let all_compliant_init_containers = pod_spec
+                .init_containers
+                .unwrap_or_else(Vec::new)
+                .into_iter()
+                .all(|container| {
+                    is_selinux_compliant(
+                        &container.security_context,
+                        expected_selinux_options.clone(),
+                        &pod_selinux_options,
+                    )
+                });
+
+            let all_compliant_ephemeral_containers = pod_spec
+                .ephemeral_containers
+                .unwrap_or_else(Vec::new)
+                .into_iter()
+                .all(|container| {
+                    is_selinux_compliant(
+                        &container.security_context,
+                        expected_selinux_options.clone(),
+                        &pod_selinux_options,
+                    )
+                });
+
+            if !all_compliant_containers
+                || !all_compliant_init_containers
+                || !all_compliant_ephemeral_containers
+            {
+                return Ok(PolicyResponse::Reject(
+                    "SELinux validation failed".to_string(),
+                ));
```

Comment:
nitpicking, it would be nice to tell the user which container is not respecting SELinux constraints

Reply:
I always find it a balance worth keeping, depending on what. On security focused policies like this, they could also give the attacker more information. For compliant targeted policies I agree, for security based ones I am not so sure about being explicit.

Reply:
I get it. I would propose, if you really want to go the extra mile, to just tell which container(s) is causing the validation to fail. The error message can stay generic, to not give away to a potential attacker the information about what are the SELinux settings that are expected.

Reply:
That's still valuable information IMO. I could create a pod with different combinations of SELinux labels and see which ones are rejected. That way I can learn about what the policy is accepting/rejecting by trying combinations.

If we reject the pod as a whole without specifying containers this task gets harder for the attacker.

I can list the containers if you think the UX vs security is worth it. I am not sure about it to be honest.